src/NuGet.Core/NuGet.Packaging/Signing/SignatureUtility.cs

```diff
@@ -0,0 +1,487 @@
+
+            if (chain == null || chain.Count == 0)
+            {
+                throw new SignatureException(errors.ChainBuildingFailed, Strings.CertificateChainBuildFailed);
+            }
+
+            return chain;
+        }
+
+        private static bool IsMatch(
+            X509Certificate2 certificate,
+            EssCertIdV2 essCertIdV2,
+            Errors errors,
+            bool isIssuerSerialRequired)
+        {
+            if (isIssuerSerialRequired)
+            {
+                if (essCertIdV2.IssuerSerial == null ||
+                    essCertIdV2.IssuerSerial.GeneralNames.Count == 0)
+                {
+                    throw new SignatureException(errors.InvalidSignature, errors.InvalidSignatureString);
+                }
+            }
+
+            if (essCertIdV2.IssuerSerial != null)
+            {
+                if (!AreSerialNumbersEqual(essCertIdV2.IssuerSerial, certificate))
+                {
+                    return false;
+                }
+
+                var generalName = essCertIdV2.IssuerSerial.GeneralNames.FirstOrDefault();
+
+                if (generalName != null &&
+                    generalName.DirectoryName != null &&
+                    generalName.DirectoryName.Name != certificate.IssuerName.Name)
+                {
+                    return false;
+                }
+            }
+
+            var hashAlgorithmName = CryptoHashUtility.OidToHashAlgorithmName(essCertIdV2.HashAlgorithm.Algorithm.Value);
+            var actualHash = CertificateUtility.GetHash(certificate, hashAlgorithmName);
+
+            return essCertIdV2.CertificateHash.SequenceEqual(actualHash);
+        }
+
+        private static bool IsMatch(X509Certificate2 certificate, EssCertId essCertId)
+        {
+            if (essCertId.IssuerSerial != null)
+            {
+                if (!AreSerialNumbersEqual(essCertId.IssuerSerial, certificate))
+                {
+                    return false;
+                }
+
+                var generalName = essCertId.IssuerSerial.GeneralNames.FirstOrDefault();
+
+                if (generalName != null &&
+                    generalName.DirectoryName != null &&
+                    generalName.DirectoryName.Name != certificate.IssuerName.Name)
+                {
+                    return false;
+                }
+            }
+
+            byte[] actualHash;
+
+            using (var hashAlgorithm = CryptoHashUtility.GetSha1HashProvider())
+            {
+                actualHash = hashAlgorithm.ComputeHash(certificate.RawData);
+            }
+
+            return essCertId.CertificateHash.SequenceEqual(actualHash);
+        }
+
+        private static bool AreSerialNumbersEqual(IssuerSerial issuerSerial, X509Certificate2 certificate)
+        {
+            var certificateSerialNumber = certificate.GetSerialNumber();
+
+            return issuerSerial.SerialNumber.SequenceEqual(certificateSerialNumber);
```

Comment:
GetSerialNumber is little endian.  The serial number in the ESSCertId(V2) should be big endian.

Reply:
Apparently your IssuerSerial type stores Little Endian.  Nevermind.